### PR TITLE
Added Sidekiq metrics to System Grafana Dashboard

### PR DESCRIPTION
**Pull request by srijitamukherjee**
Screenshot from 2020-09-08 13-49-56

Screenshot from 2020-09-08 13-50-01

Closes: https://issues.redhat.com/browse/THREESCALE-4743

Signed-off-by: Srijita Mukherjee <srmukher@redhat.com>

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -251,6 +251,7 @@ github.com/globalsign/mgo v0.0.0-20180905125535-1ca0a4f7cbcb/go.mod h1:xkRDCp4j0
 github.com/globalsign/mgo v0.0.0-20181015135952-eeefdecb41b8 h1:DujepqpGd1hyOd7aW59XpK7Qymp8iy83xq74fLr21is=
 github.com/globalsign/mgo v0.0.0-20181015135952-eeefdecb41b8/go.mod h1:xkRDCp4j0OGD1HRkm4kmhM+pmpv3AKq5SU7GMg4oO/Q=
 github.com/go-acme/lego v2.5.0+incompatible/go.mod h1:yzMNe9CasVUhkquNvti5nAtPmG94USbYxYrZfTkIn0M=
+github.com/go-bindata/go-bindata v3.1.1+incompatible h1:tR4f0e4VTO7LK6B2YWyAoVEzG9ByG1wrXB4TL9+jiYg=
```
Comment (eguzki) [inline go.sum:254]: if `go.mod` does not change, then `go.sum` should not change. Try to revert those changes and see if you can compile and run `make assets` without changes in go.sum.

Comment (srijitamukherjee) [reply go.sum:254]: Even on the master, without my changes,` make assets` modifies 

   modified:   go.sum
   modified:   pkg/assets/bindata.go

Comment (eguzki) [reply go.sum:254]: try to revert changes in go.sum. 

pkg/assets/bindata.go will change every time you run `make assets`. Only the first time includes required changes. After that, only timestamps are changed. Not a big deal. But if you change a bit of the template, do not forget to run `make assets` once again.

File: pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl [hunk 2]
```diff
@@ -69,21 +71,29 @@
       "steppedLine": false,
       "targets": [
         {
-          "expr": "sidekiq_job_runtime_seconds_count{namespace=\"$namespace\"}",
+          "expr": "sum(rate(sidekiq_jobs_success_total[1m]))",
```
Comment (eguzki) [inline pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl:74]: All expression should be "namespaced". Add `namespace` filter, should be something like this:
```
sum(rate(sidekiq_jobs_success_total{namespace='$namespace'}[1m]))
```

File: pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl [hunk 3]
```diff
@@ -69,21 +71,29 @@
...
           "format": "time_series",
-          "intervalFactor": 1,
-          "legendFormat": "{{`{{queue}}`}}-{{`{{worker}}`}}",
+          "hide": false,
+          "intervalFactor": 2,
+          "legendFormat": "Successful jobs",
           "refId": "A"
+        },
+        {
+          "expr": "sum(rate(sidekiq_jobs_failed_total[1m]))",
```
Comment (eguzki) [inline pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl:82]: namespaced

File: pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl [hunk 4]
```diff
@@ -117,438 +128,80 @@
...
       "renderer": "flot",
       "seriesOverrides": [],
       "spaceLength": 10,
-      "stack": false,
+      "stack": true,
       "steppedLine": false,
       "targets": [
         {
-          "expr": "kube_replicationcontroller_spec_replicas{namespace='$namespace',replicationcontroller=~'system-.*'}",
+          "expr": "sum(rate(sidekiq_jobs_success_total[1m])) by (queue)",
```
Comment (eguzki) [inline pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl:185]: namespaced

File: pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl [hunk 5]
```diff
@@ -117,438 +128,80 @@
...
-          "legendFormat": "{{`{{replicationcontroller}}`}}-total-pods",
-          "legendLink": null,
-          "refId": "A",
-          "step": 10
+          "legendFormat": "Processed in {{`{{queue}}`}}",
+          "refId": "A"
         },
         {
-          "expr": "kube_replicationcontroller_status_ready_replicas{namespace='$namespace',replicationcontroller=~'system-.*'}",
+          "expr": "sum(rate(sidekiq_jobs_failed_total[1m])) by (queue)",
```
Comment (eguzki) [inline pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl:193]: namespaced

File: pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl [hunk 6]
```diff
@@ -592,55 +245,70 @@
...
       "renderer": "flot",
       "seriesOverrides": [],
       "spaceLength": 10,
-      "stack": false,
+      "stack": true,
       "steppedLine": false,
       "targets": [
         {
-          "expr": "sum(delta(kube_pod_container_status_restarts_total{namespace='$namespace',pod=~'system-.*'}[5m])) by (pod)",
+          "expr": "sum(rate(sidekiq_jobs_success_total[1m])) by (worker)",
```
Comment (eguzki) [inline pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl:296]: namespaced

**Review by eguzki — CHANGES_REQUESTED** [review]
Looking good.

All expr should be "namespaced". I left a comment for some of the expr. Make sure, all expr are "namespaced".

I am not going to review the actual changes of the dashboard. If that is the dashboard you want, let it be. 

Rebase on top of current master head and it will be ready to be merged.

File: pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl [hunk 7]
```diff
@@ -117,438 +128,80 @@
         "alignLevel": null
       }
     },
-    {
```
Comment (hallelujah) [inline pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl:120]: So this should be kept. Do not remove the operations metrics

**Review by hallelujah — CHANGES_REQUESTED** [review]
We should not remove the metrics already defined.

File: pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl (hunk 7 above)
Comment (hallelujah) [reply pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl:120]: This is also an old version of pod resources
Lets migrate it to the new version, copy the zync or backend dashboard and add the application metrics.

Comment (srijitamukherjee) [reply pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl]: If I understand this correctly, we want to the pods resource consumption(CPU-Usage/Quota, Memory-Usage/Quota, Network-Usage) distributed per Deployment Configs.

File: pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl [hunk 8]
```diff
@@ -45,6 +48,556 @@
...
+      "pointradius": 2,
+      "points": false,
+      "renderer": "flot",
+      "seriesOverrides": [],
+      "spaceLength": 10,
+      "stack": true,
+      "steppedLine": false,
+      "targets": [
+        {
+          "expr": "max(sidekiq_jobs_waiting_count) by (queue)",
```
Comment (raelga) [inline pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl:363]:
```suggestion
          "expr": "max(sidekiq_jobs_waiting_count{namespace='$namespace'}) by (queue)",
```

File: pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl [hunk 9]
```diff
@@ -45,6 +48,556 @@
...
+      "pointradius": 2,
+      "points": false,
+      "renderer": "flot",
+      "seriesOverrides": [],
+      "spaceLength": 10,
+      "stack": false,
+      "steppedLine": false,
+      "targets": [
+        {
+          "expr": "max(sidekiq_queue_latency) by (queue)",
```
Comment (raelga) [inline pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl:453]:
```suggestion
          "expr": "max(sidekiq_queue_latency{namespace='$namespace'}) by (queue)",
```

File: pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl [hunk 10]
```diff
@@ -45,6 +48,556 @@
...
+      "pointradius": 2,
+      "points": false,
+      "renderer": "flot",
+      "seriesOverrides": [],
+      "spaceLength": 10,
+      "stack": false,
+      "steppedLine": false,
+      "targets": [
+        {
+          "expr": "histogram_quantile(0.95, sum(rate(sidekiq_job_runtime_seconds_bucket[1m])) by (le)) ",
```
Comment (raelga) [inline pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl:539]:
```suggestion
          "expr": "histogram_quantile(0.95, sum(rate(sidekiq_job_runtime_seconds_bucket{namespace='$namespace'}[1m])) by (le)) ",
```

File: pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl [hunk 11]
```diff
@@ -69,18 +622,107 @@
       "steppedLine": false,
       "targets": [
         {
-          "expr": "sidekiq_job_runtime_seconds_count{namespace=\"$namespace\"}",
+          "expr": "histogram_quantile(0.95, sum(rate(sidekiq_job_runtime_seconds_bucket[1m])) by (worker, le)) ",
```
Comment (raelga) [inline pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl:625]:
```suggestion
          "expr": "histogram_quantile(0.95, sum(rate(sidekiq_job_runtime_seconds_bucket{namespace='$namespace'}[1m])) by (worker, le)) ",
```

File: pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl [hunk 12]
```diff
@@ -69,18 +622,107 @@
...
+      "pointradius": 2,
+      "points": false,
+      "renderer": "flot",
+      "seriesOverrides": [],
+      "spaceLength": 10,
+      "stack": true,
+      "steppedLine": false,
+      "targets": [
+        {
+          "expr": "sum(rate(sidekiq_jobs_enqueued_total[1m])) by (queue)",
```
Comment (raelga) [inline pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl:714]:
```suggestion
          "expr": "sum(rate(sidekiq_jobs_enqueued_total{namespace='$namespace'}[1m])) by (queue)",
```

File: pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl [hunk 13]
```diff
@@ -95,13 +737,102 @@
...
+      "pointradius": 2,
+      "points": false,
+      "renderer": "flot",
+      "seriesOverrides": [],
+      "spaceLength": 10,
+      "stack": true,
+      "steppedLine": false,
+      "targets": [
+        {
+          "expr": "max(sidekiq_jobs_retry_count)",
```
Comment (raelga) [inline pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl:803]:
```suggestion
          "expr": "max(sidekiq_jobs_retry_count{namespace='$namespace'})",
```

**Review by raelga — COMMENTED** [review]
Thanks for the improvement @srijitamukherjee!

The review adds some missing namespace filters and also improves the deploymentConfig selector (it was getting `system-app` and `system-app-1`).

**Review by raelga — CHANGES_REQUESTED** [review]
The previous review require some changes.

File: pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl (hunk 13 above)
Comment (srijitamukherjee) [reply pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl:803]: I'll make the changes for all mentioned above.

File: pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl [hunk 14]
```diff
@@ -1601,6 +2346,27 @@
...
+        "definition": "label_values(kube_pod_info{namespace='$namespace',pod=~'system-.*'}, pod)",
+        "hide": 0,
+        "includeAll": true,
+        "label": "deploymentConfig",
+        "multi": false,
+        "name": "deploymentConfig",
+        "options": [],
+        "query": "label_values(kube_pod_info{namespace='$namespace',pod=~'system-.*'}, pod)",
+        "refresh": 1,
+        "regex": "/(.*)-[a-z0-9]+-[a-z0-9]+/",
```
Comment (raelga) [inline pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl:2362]:
```suggestion
        "regex": "/([a-z]+-[a-z]+)-[0-9]*/",
```
Screenshot 2020-10-14 at 11 56 51
Screenshot 2020-10-14 at 11 57 17

**Review by raelga — APPROVED** [review]
Thanks @srijitamukherjee for addressing the review!

**Pull request merged**